src/video/video.c

```diff
@@ -1,306 +1,174 @@
 
-#include "utils/allocator.h"
 #include "utils/log.h"
-#include "video/opengl/object_array.h"
-#include "video/opengl/remaps.h"
-#include "video/opengl/render_target.h"
-#include "video/opengl/shaders.h"
-#include "video/opengl/shared.h"
-#include "video/opengl/texture_atlas.h"
-#include "video/sdl_window.h"
-#include "video/vga_state.h"
+#include "video/renderers/renderer.h"
 #include "video/video.h"
 
-typedef struct video_state {
-    SDL_Window *window;
-    SDL_GLContext *gl_context;
-    texture_atlas *atlas;
-    object_array *objects;
-    shared *shared;
-    render_target *target;
-    remaps *remaps;
+// If-def the includes here
+#include "video/renderers/opengl3/gl3_renderer.h"
+#ifdef ENABLE_NULL_RENDERER
+#include "video/renderers/null/null_renderer.h"
+#endif
+
+typedef void (*renderer_init)(renderer *renderer);
+typedef struct renderer_status {
+    renderer_init init;
+    bool is_available;
+    const char *name;
+    const char *description;
+} renderer_status;
+
+// If-def the renderers here. Most preferred renderers at the top.
+static renderer_status renderer_statuses[] = {
+    {gl3_renderer_init,  false, NULL, NULL},
+#ifdef ENABLE_NULL_RENDERER
+    {null_renderer_init, false, NULL, NULL},
+#endif
+};
+
+static const int renderer_count = sizeof(renderer_statuses) / sizeof(renderer_status);
+static renderer current_renderer;
 
-    GLuint palette_prog_id;
-    GLuint rgba_prog_id;
-
-    object_array_blend_mode current_blend_mode;
-
-    int viewport_w;
-    int viewport_h;
-
-    int screen_w;
-    int screen_h;
-    bool fullscreen;
-    bool vsync;
-
-    bool draw_atlas;
-    int target_move_x;
-    int target_move_y;
-
-    video_screenshot_signal screenshot_cb;
-} video_state;
-
-#define TEX_UNIT_ATLAS 0
-#define TEX_UNIT_FBO 1
-#define TEX_UNIT_REMAPS 2
-
-#define PAL_BLOCK_BINDING 0
-
-static video_state g_video_state;
-
-int video_init(int window_w, int window_h, bool fullscreen, bool vsync) {
-    g_video_state.screen_w = window_w;
-    g_video_state.screen_h = window_h;
-    g_video_state.fullscreen = fullscreen;
-    g_video_state.vsync = vsync;
-    g_video_state.target_move_x = 0;
-    g_video_state.target_move_y = 0;
-    g_video_state.current_blend_mode = MODE_SET;
-
-    if(!create_window(&g_video_state.window, window_w, window_h, fullscreen)) {
-        goto error_0;
-    }
-    if(!create_gl_context(&g_video_state.gl_context, g_video_state.window)) {
-        goto error_1;
-    }
-    if(!set_vsync(g_video_state.vsync)) {
-        goto error_2;
-    }
-    if(!create_program(&g_video_state.palette_prog_id, "palette.vert", "palette.frag")) {
-        goto error_2;
-    }
-    if(!create_program(&g_video_state.rgba_prog_id, "rgba.vert", "rgba.frag")) {
-        goto error_3;
+/**
+ * This is run at start to hunt the available renderers.
+ */
+void video_scan_renderers(void) {
+    renderer tmp;
+    bool available;
+    for(int i = 0; i < renderer_count; i++) {
+        renderer_statuses[i].init(&tmp);
```

Comment:
renderer init without close leaks memory

Reply:
There is no corresponding close for gl3_renderer_init(), since it does not reserve memory

Reply:
Reorganized this part to be clearer.